File: backend/poseidon/core/identity.py
"""Phase 9 Task 1 (doc 05 section 2, decision D22): the IdentityProvider
seam every downstream module threads a real caller identity through, and
the one provider this task ships -- ``DisabledProvider``, the default mode
every existing test/env already boots under.

**Provider-blindness (D22).** Everything BELOW the seam -- ``api/live_chat.
py``, ``core/chat/orchestrator.py``, ``core/runlog.py``'s writer calls,
``core/skills/context.py``'s ``SkillContext.user`` -- reads exactly one
already-resolved :class:`UserContext` and never branches on
``settings.identity_mode`` or imports anything from this module except that
one type. The mode switch lives in exactly one place: :func:`resolve_
provider`, called once at app construction (``api/app.py``'s ``create_app``)
so a misconfigured or not-yet-implemented mode fails at BOOT, the same
"no half-configured server ever accepts traffic" discipline ``core/config.
py``'s own module docstring already states for a malformed ``Settings``
value.

**Providers never import FastAPI.** :meth:`IdentityProvider.resolve` takes
a plain ``Mapping[str, str]`` -- a header bag with LOWERCASE keys, never a
``starlette.requests.Request`` -- so this whole module (and every provider
that will ever implement this protocol: ``Auth0Provider``/
``SpcsIngressProvider``, Phase 9 Tasks 2/3) stays free of any web-framework
import. ``api/auth.py``'s identity middleware is the ONE adapter: it
lowercases a real request's headers (``{k.lower(): v for k, v in request.
headers.items()}``) and calls ``provider.resolve(headers)``, storing the
result on ``request.state.user`` for every downstream dependency/handler to
read back. A unit test exercises a provider directly with a plain dict
built the same way -- no ASGI app, no ``Request`` construction, needed (see
``test_identity_providers.py``).

**Why a mapping, not a request.** Every mode this plan ships resolves
identity from exactly one HTTP header (``X-Dev-User`` here; ``Authorization``
for Auth0; ``Sf-Context-Current-User`` for the SPCS ingress trust boundary)
-- a mapping is the smallest honest shape that need, and it is what makes a
provider testable in complete isolation from the web framework carrying it.

**What this task ships vs. what it does not.** :class:`UserContext`,
:class:`AuthError`, the :class:`IdentityProvider` protocol, :class:`
DisabledProvider` and :func:`resolve_provider` all ship now. ``AuthError``
is defined here because the PROTOCOL needs a pinned failure shape to
document, but no provider in this task ever raises it: ``DisabledProvider``
never rejects a request (an invalid ``X-Dev-User`` header is IGNORED, not
an error -- see that class's own docstring), so the AuthError-to-HTTP-
response mapping (401/403 RFC-7807 problem shapes) is deliberately left for
Phase 9 Task 2, the first task whose provider (``Auth0Provider``) can
actually raise it. Building that mapping now, against no real raiser and no
pinned problem shape yet, would be speculative code this task cannot prove
with a test.
"""

import re
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Protocol

from poseidon.core.config import Settings


@dataclass(frozen=True)
class UserContext:
    """The one identity shape every mode (``disabled``/``auth0``/
    ``spcs_ingress``) resolves down to (doc 05 section 2). ``sub`` is always
    provider-prefixed (``dev|...`` here; ``auth0|...``/``sf|...`` for the
    other two modes) so a sub is globally unambiguous about which identity
    system minted it, even after Phase 10 starts keying persisted rows off
    it. Frozen, like every other value type in this codebase -- a caller
    cannot rewrite the identity a request resolved to out from under
    whatever already read it.
    """

    sub: str
    email: str | None
    name: str | None
    roles: tuple[str, ...]


class AuthError(Exception):
    """Raised by :meth:`IdentityProvider.resolve` when a request cannot be
    resolved to a :class:`UserContext` -- an invalid/expired/tampered
    credential (Auth0), or a trust violation (SPCS ingress outside the
    trusted edge). Carries the same three fields ``core.skills.result.
    problem()`` renders into an RFC-7807 body, so the api layer can map one
    straight onto the other without this module (or any provider) importing
    FastAPI or knowing anything about HTTP response shapes -- see the
    module docstring's "providers never import FastAPI" seam.

    No provider THIS task ships (:class:`DisabledProvider`) ever raises
    this; see the module docstring's "What this task ships" for why the
    api-layer mapping is deliberately deferred to Phase 9 Task 2, the first
    task with a provider that can actually raise it.
    """

    def __init__(self, status: int, title: str, detail: str) -> None:
        super().__init__(detail)
        self.status = status
        self.title = title
        self.detail = detail


class IdentityProvider(Protocol):
    """The one seam every identity mode implements. See the module
    docstring for the full provider-blindness/no-FastAPI-import contract
    this protocol exists to enforce."""

    def resolve(self, headers: Mapping[str, str]) -> UserContext:
        """Resolve one request's identity from its (lowercase-keyed)
        headers. Returns a :class:`UserContext`, or raises :class:`AuthError`
        for a request that cannot be resolved -- see that class's own
        docstring for which modes actually take the error path."""
        ...


# ---------------------------------------------------------------------------
# disabled mode (the default): fixed identity + the X-Dev-User act-as header
# ---------------------------------------------------------------------------

# The plan's own pinned fixed context (Global Constraints), generalized from
# Phase 6-8's bare DEV_USER_SUB = "dev|local" string constant into the full
# UserContext shape this seam carries everywhere now. The sub value itself
# is unchanged -- every existing row/test that pinned "dev|local" stays
# coherent (Task 1's own contract pin) -- it now arrives THROUGH
# DisabledProvider rather than as a constant orchestrator.py reached for
# directly.
DISABLED_DEFAULT_USER = UserContext(
    sub="dev|local", email="dev@local", name="Dev User", roles=("Poseidon:Sales",)
)

# wfs's own act-as convention for multi-user local testing (Global
# Constraints): a browser tab (or curl call) can pose as a named dev user
# without any real login flow existing yet. Lowercase -- the header-mapping
# contract this whole seam relies on (see the module docstring's "providers
# never import FastAPI"): api/auth.py's middleware lowercases every header
# name before calling in, so this key is always found regardless of the
# case the client actually sent ("X-Dev-User", "x-dev-user", ... all land
# here identically).
_ACT_AS_HEADER = "x-dev-user"

# Sanitize rule (Task 1's own pinned contract): casefold the raw header
# value FIRST, then require the WHOLE string to match this character class,
# 1-64 of them. Anything that does not match means "ignore the header", not
# "reject the request" -- disabled mode never 4xxs on a bad act-as value; it
# just falls back to the fixed default, since the entire point of this mode
# is "never blocks local dev, ever". fullmatch (not search/match) is what
# makes this a WHOLE-STRING contract: "alice!" must fail even though its
# "alice" prefix alone would match.
_ACT_AS_PATTERN = re.compile(r"[a-z0-9_-]{1,64}")


class DisabledProvider:
    """``identity_mode="disabled"`` -- the default, and the ONLY mode this
    task implements (Global Constraints: "every existing test/env boots
    unchanged"). Every request resolves to :data:`DISABLED_DEFAULT_USER`,
    unless the caller sends a valid ``X-Dev-User`` header, in which case a
    sanitized value ``X`` resolves to sub ``"dev|X"`` instead (see the
    module-level constants above for the exact sanitize rule).

    ``email``/``name`` for an act-as identity are NOT pinned by the plan --
    only the sub shape (``"dev|alice"``) is a contract pin. Derived here as
    ``"{X}@local"``/``X`` (a judgment call, disclosed in this task's report)
    so a UI rendering ``{email, name, roles}`` for an act-as user shows
    something obviously synthetic and DISTINCT per act-as identity, rather
    than reusing ``"dev@local"``/``"Dev User"`` verbatim for every one of
    them -- which would make every ``dev|X`` sub look like the exact same
    person everywhere except the one field a UI is least likely to surface
    prominently. ``roles`` stays the fixed ``("Poseidon:Sales",)`` for every
    act-as identity regardless: disabled mode has no role system of its own
    to differentiate against, and ``Poseidon:Sales`` is the one role every
    route this phase gates actually requires.
    """

    def resolve(self, headers: Mapping[str, str]) -> UserContext:
        raw = headers.get(_ACT_AS_HEADER)
        if raw is None:
            return DISABLED_DEFAULT_USER
        candidate = raw.casefold()
        if _ACT_AS_PATTERN.fullmatch(candidate) is None:
            # Invalid -- pinned as "ignore the header", never a rejection;
            # the fixed default answers instead, identically to no header.
            return DISABLED_DEFAULT_USER
        return UserContext(
            sub=f"dev|{candidate}",
            email=f"{candidate}@local",
            name=candidate,
            roles=("Poseidon:Sales",),
        )


def resolve_provider(settings: Settings) -> IdentityProvider:
    """Select the :class:`IdentityProvider` named by ``settings.
    identity_mode``. Called once, at app construction (``api/app.py``'s
    ``create_app``) -- never lazily, on first request -- so a misconfigured
    or not-yet-implemented mode fails at BOOT rather than serving even one
    request under the wrong identity story.

    ``"disabled"``/``"auth0"`` resolve today. ``"spcs_ingress"`` is a real
    ``Settings.identity_mode`` value (the ``Literal`` already accepts all
    three -- doc 05 section 2's full design), but its provider ships in
    Phase 9 Task 3; selecting it before then hits the exact same fail-fast
    branch a genuinely unrecognized string would (impossible to reach
    through ``Settings`` alone, whose ``Literal`` already rejects anything
    else at config-load time -- this is the defensive belt for that
    already-enforced suspenders, doubling as the honest answer for a mode
    this task has not implemented yet).
    """
    if settings.identity_mode == "disabled":
        return DisabledProvider()
    if settings.identity_mode == "auth0":
        # Imported here, not at module level: identity_auth0.py itself
        # imports AuthError/UserContext FROM this module, so a top-level
        # import here would be a genuine circular import (this module
        # would try to import identity_auth0 before its own AuthError/
        # UserContext names exist yet for identity_auth0 to import back).
        # Deferring to call time -- long after both modules are fully
        # loaded -- sidesteps that entirely. Phase 9 Task 3 will add the
        # identical one-line branch + deferred import for spcs_ingress.
        from poseidon.core.identity_auth0 import Auth0Provider

        return Auth0Provider(settings)
    raise RuntimeError(
        f"identity_mode={settings.identity_mode!r} has no IdentityProvider implemented yet"
    )


__all__ = [
    "AuthError",
    "DISABLED_DEFAULT_USER",
    "DisabledProvider",
    "IdentityProvider",
    "UserContext",
    "resolve_provider",
]

File: web/src/lib/identity.ts
/**
 * Identity resolution, ported from the Python providers this app still runs
 * behind: `backend/poseidon/core/identity.py` (`DisabledProvider`) and
 * `backend/poseidon/core/identity_spcs.py` (`SpcsIngressProvider`).
 *
 * **Why byte-identical matters.** `sub` is the key every persisted row and
 * every row-level-security policy is written against. A sub this module mints
 * even one character differently from the sub Python mints for the same
 * request does not fail loudly -- the user silently loses their history and
 * RLS silently stops matching. Every value below is therefore a copy of a
 * pinned Python value, cited to the line that pins it, not a re-derivation.
 *
 * **One accepted divergence.** Python normalises with `str.casefold()`;
 * JavaScript has only `String.prototype.toLowerCase()`. They differ on some
 * non-ASCII input (`"ẞ"` casefolds to `"ss"`, which passes the character
 * class below, but lower-cases to `"ß"`, which does not). Every difference
 * found runs in that direction: TypeScript rejects a value Python would have
 * accepted, so the divergence fails CLOSED -- a request is refused, never
 * resolved to the wrong sub. Deliberately not closed with a case-folding
 * dependency.
 */

export type Identity = {
  sub: string;
  email: string | null;
  name: string | null;
  roles: string[];
};

/**
 * A credential/trust failure: the caller did not present an identity this app
 * can accept.
 *
 * Carries Python's `(status, title, detail)` triple (`identity.py:99-103`) and,
 * exactly as Python does, hands `detail` to `Error`'s message -- so
 * `err.message` here is the string `str(exc)` gives there.
 *
 * The triple is what lets the ONE place that renders this (`proxy.ts`) emit the
 * same RFC-7807 body every other failure in this codebase renders through
 * `problem()` (`core/skills/result.py:72-79`, reached via `api/auth.py:179`),
 * instead of hard-coding a status and a title of its own. A second, hand-rolled
 * problem shape is precisely what `auth_error_response`'s docstring exists to
 * forbid on the Python side.
 */
export class AuthError extends Error {
  readonly status: number;
  readonly title: string;
  readonly detail: string;

  constructor(status: number, title: string, detail: string) {
    super(detail);
    this.name = "AuthError";
    this.status = status;
    this.title = title;
    this.detail = detail;
  }
}

/**
 * A CONFIGURATION fault: the operator's settings are wrong, and no credential
 * the caller could present would change the outcome.
 *
 * Deliberately NOT a subclass of `AuthError`, because Python keeps
 * exactly this split -- `RuntimeError` for a mode that is misconfigured or
 * unimplemented (`identity.py:301-303`, `identity_spcs.py:131-137`), and
 * `AuthError` only for a caller's credential (`api/auth.py:21`). Collapsing the
 * two renders an operator's mistake as a 401, which tells the user to log in
 * again for something only an operator can fix and buries the real fault in a
 * wall of authentication failures. `proxy.ts` renders this as a 500 instead.
 *
 * Python's equivalents fire at BOOT (`resolve_provider` builds the provider
 * once), so a misconfigured server there never accepts traffic at all. This
 * port has no boot-time seam yet, so it fails per request. Both fail closed and
 * neither ever trusts an untrustworthy header; the Python one is louder. If a
 * later task adds boot-time config validation, this class is what it should
 * raise there.
 */
export class IdentityConfigError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "IdentityConfigError";
  }
}

/**
 * The one role every route this app gates requires. Python carries two
 * deliberate copies of this literal (`identity.py`'s `DISABLED_DEFAULT_USER`
 * and `identity_spcs.py`'s `_SALES_ROLE`) because its providers may not import
 * the api layer; this port needs only one.
 */
const SALES_ROLE = "Poseidon:Sales";

/**
 * Mirrors `DISABLED_DEFAULT_USER` (identity.py:130) field for field.
 *
 * A function, not a shared constant: Python's `UserContext` is a frozen
 * dataclass holding a `tuple`, so no caller can mutate one resolved identity
 * into another. TypeScript has no cheap equivalent, and handing every caller
 * the same object -- and, worse, the same `roles` array -- means one stray
 * `identity.roles.push(...)` grants that role to every subsequent request for
 * the life of the process. Minting a fresh object per call is the honest port.
 */
function disabledDefaultUser(): Identity {
  return { sub: "dev|local", email: "dev@local", name: "Dev User", roles: [SALES_ROLE] };
}

/**
 * Mirrors `DisabledProvider.resolve`'s act-as branch (identity.py:214-219).
 *
 * `email`/`name` are DERIVED PER USER (`"{X}@local"` / `X`), not inherited
 * from the fixed default above. That is Python's actual behaviour, and its
 * reason (that provider's own docstring): reusing `"dev@local"`/`"Dev User"`
 * for every act-as identity would make every distinct `dev|X` sub render as
 * the same person everywhere a UI shows a name. `roles` stays the fixed sales
 * role -- disabled mode has no role system of its own to differentiate with.
 */
function disabledActAsUser(username: string): Identity {
  return {
    sub: `dev|${username}`,
    email: `${username}@local`,
    name: username,
    roles: [SALES_ROLE],
  };
}

/** Act-as header, lowercase -- see `sanitizeUsername` on header casing. */
const ACT_AS_HEADER = "x-dev-user";

/**
 * The header the SPCS platform ingress edge attaches after authenticating the
 * visitor as a Snowflake user (`identity_spcs.py`'s `_SF_CONTEXT_HEADER`).
 */
const SF_CONTEXT_HEADER = "sf-context-current-user";

/**
 * The title and detail of the one 401 this module raises, character for
 * character from `identity_spcs.py:113-114` (`_MISSING_HEADER_TITLE` /
 * `_MISSING_HEADER_DETAIL`). Both halves reach the wire through the RFC-7807
 * body, so both are part of the same parity contract as the subs themselves.
 */
const MISSING_HEADER_TITLE = "missing spcs identity header";
const MISSING_HEADER_DETAIL = "no valid Sf-Context-Current-User header";

/** `identity_spcs.py`'s `_ALLOW_ALL`: "everyone the edge vouches for". */
const ALLOW_ALL = "*";

/**
 * Port of `sanitize_username` (identity.py:155) -- the ONE rule this codebase
 * applies to any operator- or platform-supplied username-shaped header, shared
 * by both providers there and so by both modes here.
 *
 * Python: `_ACT_AS_PATTERN = re.compile(r"[a-z0-9_-]{1,64}")` applied with
 * `.fullmatch()` to the casefolded value. Two things that character class
 * excludes are easy to add back by accident: there is NO dot (so `"first.last"`
 * is not a username), and there is a 1-64 length cap. `fullmatch` -- anchored
 * here as `^...$` -- is why `"alice!"` is rejected wholesale rather than
 * truncated to its matching `"alice"` prefix, which would silently resolve two
 * different callers to one sub.
 *
 * Header NAMES need no normalising: `Headers.get` is case-insensitive, the
 * same guarantee Python gets from `api/app.py` lowercasing every header name
 * before a provider sees it.
 */
const SAFE_NAME = /^[a-z0-9_-]{1,64}$/;

/**
 * Returns the sanitised username, or null when it does not match. Exported
 * for the same reason Python names it in its own `__all__`: it is the one
 * shared rule, and the next provider to read a username-shaped header should
 * call it rather than re-derive the character class.
 */
export function sanitizeUsername(raw: string): string | null {
  const candidate = raw.toLowerCase();
  return SAFE_NAME.test(candidate) ? candidate : null;
}

/**
 * Reads the sales allowlist the way `Settings.spcs_sales_users` is built and
 * then consumed: `config.py`'s `split_spcs_sales_users` splits on commas and
 * strips each entry (dropping empties), and `SpcsIngressProvider.__init__`
 * casefolds the result -- so neither an operator's spacing nor their casing in
 * `SPCS_SALES_USERS` decides who gets the role. Unset means an EMPTY list, the
 * fail-closed default `config.py` picks on purpose: nobody holds the sales
 * role until an operator names someone (or `"*"`).
 */
function salesAllowlist(): Set<string> {
  return new Set(
    (process.env.SPCS_SALES_USERS ?? "")
      .split(",")
      .map((name) => name.trim().toLowerCase())
      .filter(Boolean),
  );
}

export function resolveIdentity(
  mode: "disabled" | "spcs_ingress" | "auth0",
  deployMode: string,
  headers: Headers,
): Identity {
  if (mode === "disabled") {
    const raw = headers.get(ACT_AS_HEADER);
    const username = raw === null ? null : sanitizeUsername(raw);
    if (username === null) {
      // An absent OR invalid act-as header is IGNORED, never rejected. The
      // whole point of this mode is that it never refuses a request, so the
      // fixed default answers for both -- identically (identity.py:211-213).
      return disabledDefaultUser();
    }
    return disabledActAsUser(username);
  }

  if (mode === "spcs_ingress") {
    // The header below is unsigned and unverifiable. It is trustworthy ONLY
    // because the SPCS platform edge is the one component that can attach it
    // to a request reaching this app. Anywhere else -- a laptop, an EC2 box --
    // any caller could set it to any value and this code could not tell the
    // difference. Python enforces this in `SpcsIngressProvider.__init__`
    // (identity_spcs.py:131-137), so it fails at BOOT; see
    // `IdentityConfigError` for why this is that class and not an `AuthError`,
    // and for what this port gives up by failing per request instead.
    if (deployMode !== "spcs") {
      throw new IdentityConfigError(
        "spcs identity header is not trusted outside spcs deploy mode",
      );
    }
    const raw = headers.get(SF_CONTEXT_HEADER);
    const username = raw === null ? null : sanitizeUsername(raw);
    // A present-but-malformed header raises the SAME error as an absent one:
    // both mean the trusted edge did not deliver what it guarantees, and there
    // is no default identity to fall back to in the mode a real deployment
    // runs under. Python has no separate "malformed" bucket here either
    // (identity_spcs.py's `resolve`).
    if (username === null) {
      throw new AuthError(401, MISSING_HEADER_TITLE, MISSING_HEADER_DETAIL);
    }
    // Roles are ALLOWLIST-GATED, not granted to everyone the platform
    // authenticates -- identity_spcs.py:156,
    // `roles = (_SALES_ROLE,) if self._is_allowed(candidate) else ()`.
    // A user the platform authenticated but who is NOT on the list resolves
    // successfully with an EMPTY role list: authenticated, but not authorised.
    // `require_sales` on the backend is the one place that becomes a 403.
    const allowlist = salesAllowlist();
    const allowed = allowlist.has(ALLOW_ALL) || allowlist.has(username);
    // email/name stay null rather than reusing the username: the header
    // carries a bare username and nothing else, and a production identity path
    // should not assert a display name or email it cannot vouch for. (This is
    // exactly why disabled mode's synthetic dev identities, above, may.)
    return {
      sub: `sf|${username}`,
      email: null,
      name: null,
      roles: allowed ? [SALES_ROLE] : [],
    };
  }

  if (mode === "auth0") {
    // A config fault, not a credential one: Python HAS an auth0 provider, so
    // the only way to reach this in the port is an operator selecting a mode
    // this phase has not built. That is the same fault as the unimplemented-mode
    // tail below (identity.py:301-303's `RuntimeError`), and a caller's
    // credential is irrelevant to it.
    throw new IdentityConfigError("auth0 mode is not wired in this phase (decision M5)");
  }

  // Unreachable through the declared type, but `IDENTITY_MODE` is an
  // unvalidated env string that `proxy.ts` casts into it, so a typo really can
  // arrive here. Naming the offending value beats letting it fall into the
  // auth0 branch and reporting a mode the operator never configured -- the
  // same defensive tail `resolve_provider` keeps in identity.py:301.
  throw new IdentityConfigError(
    `identity_mode=${JSON.stringify(mode)} has no resolver implemented`,
  );
}
